crochetcraft/src/routes/tools/curveBuilder/+page.svelte

```diff
@@ -0,0 +1,198 @@
+    const interpolationOptions = enumToKeys(InterpolationMode);
+
+    type EditorSettings = {
+        drawType: DrawType;
+        interpolationMode: InterpolationMode;
+        interpolationSamples: number;
+        duplicates: number;
+    };
+
+    let scene: THREE.Scene;
+    let points: Vector3[] = [new Vector3(0, 0, 0), new Vector3(1, 1, 1)];
+    let editorSettings: EditorSettings = {
+        drawType: DrawType.Point,
+        interpolationMode: InterpolationMode.Linear,
+        interpolationSamples: 10,
+        duplicates: 0,
+    };
+    let quicklink = '';
+
+    const superGroup = new THREE.Group();
+
+    function getCurve(): THREE.Curve<Vector3> {
+        switch (editorSettings.interpolationMode) {
+            case InterpolationMode.Linear:
+                const path = new THREE.CurvePath<Vector3>();
+                for (let i = 1; i < points.length; ++i) {
+                    path.add(new THREE.LineCurve3(points[i - 1], points[i]));
+                }
+                return path;
+            case InterpolationMode.CubicSpline:
+                return new CubicSpline3D(points);
+        }
+    }
+
+    function drawCurve(path: THREE.Curve<Vector3>): THREE.Object3D {
+        switch (editorSettings.drawType) {
+            case DrawType.Point:
+                const subGroup = new THREE.Group();
+                placeDebugSpheres(path.getPoints(editorSettings.interpolationSamples), subGroup);
+                return subGroup;
+            case DrawType.Line:
+                return new THREE.Line(
+                    new THREE.BufferGeometry().setFromPoints(
+                        path.getPoints(editorSettings.interpolationSamples),
+                    ),
+                    new THREE.MeshBasicMaterial(),
+                );
+            case DrawType.Cylinder:
+                return new THREE.Mesh(
+                    new THREE.TubeGeometry(path, editorSettings.interpolationSamples, 0.5, 8),
+                    new THREE.MeshLambertMaterial(),
+                );
+        }
+    }
+    onMount(() => {
+        const path = new URLSearchParams(window.location.search).get('p');
+        if (path) {
+            try {
+                points = JSON.parse(decodeURI(path)).map(
+                    (p: { x: number; y: number; z: number }) => new Vector3(p.x, p.y, p.z),
+                );
+            } catch {
+                console.error("Invalid search params, 'points' will just contain defaults.");
+            }
+        }
+
+        // add static, angled directional light
+        const lightB = new THREE.DirectionalLight(0x000020, 100);
+        scene.add(lightB);
+        lightB.target = new THREE.Object3D().translateX(10).translateY(5).translateZ(2);
+        scene.add(lightB.target);
+
+        const lightR = new THREE.DirectionalLight(0x200000, 100);
+        scene.add(lightR);
+        lightR.target = new THREE.Object3D().translateX(-10).translateY(-5).translateZ(-2);
+        scene.add(lightR.target);
```

Comment:
Hmm, I wonder if this could be put into `ThreeCanvas` itself... (You don't have to do this, I'm just pondering.)

Reply:
Having any lights in the original ThreeCanvas was probably a mistake imo.
It probably makes more sense to have helper functions that setup lights in a scene